Output C code from harness generation tool.

NlightNFotis opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

hannes-steffenhagen-diffblue commented on regression/goto-harness/havoc-global-int-02/main.c:
```diff
@@ -3,7 +3,7 @@
 unsigned int y;
 
-unsigned int nondet_int()
+unsigned int some_random_unsigned_integer()
```
This was changed because of a `dump_c()` bug; It assumes that `nondet_*` names are available with specific meanings.

hannes-steffenhagen-diffblue commented on regression/goto-harness/havoc-global-int-02/test.desc:
```diff
@@ -1,4 +1,4 @@
-CORE
+KNOWNBUG
```
We can work around this for now by re-enabling the old behaviour for these, but we've marked them KNOWNBUG for now because we're first trying to get the call-function harness to behave like we want it to.

thk123 replied:
:no_entry_sign: As a rule - tests marked as knownbug should reference either a jira ticket or github issue that explains how they should be turned back on - otherwise we might as well delete them. Could you raise a Jira ticket to this effect and put it in the comment section of each desc file

hannes-steffenhagen-diffblue commented on src/goto-harness/common_harness_generator_options.h:
```diff
@@ -18,6 +18,7 @@ Author: Diffblue Ltd.
   "function-pointer-can-be-null"
 #define COMMON_HARNESS_GENERATOR_HAVOC_MEMBER_OPT "havoc-member"
 
+
```
🚫 Whitespace

hannes-steffenhagen-diffblue commented on src/goto-harness/function_call_harness_generator.cpp:
```diff
@@ -215,6 +215,8 @@ void function_call_harness_generatort::handle_option(
   }
 }
 
+#include <iostream>
```
🚫 debug code changes

hannes-steffenhagen-diffblue commented on src/goto-harness/function_call_harness_generator.cpp:
```diff
@@ -478,7 +480,7 @@ function_call_harness_generatort::implt::declare_arguments(
   {
     auto argument = allocate_objects.allocate_automatic_local_object(
-      parameter.type(), parameter.get_base_name());
+      remove_const(parameter.type()), parameter.get_base_name());
```
The `remove_const` changes should probably all be in one separate commit.

hannes-steffenhagen-diffblue commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -1002,7 +1021,7 @@ code_blockt recursive_initializationt::build_function_pointer_constructor(
     ++function_pointer_index;
   }
-
+  
```
stray whitespace change

NlightNFotis committed:
Add remove_const function that removes C const qualifier from type ireps

hannes-steffenhagen-diffblue committed:
Initialise const values recursively.

Previously the `dump-c` output of recursive initialisation
wouldn't work with types that have nested const things in
them, such as pointer-to-const or array with constant elements.

NlightNFotis committed:
Add missing local declaration in recursive_initialization.cpp

NlightNFotis committed:
Mark snapshot tests as KNOWNBUG

The C output doesn't quite work with the snapshot harness at the moment
(specifically, it doesn't work with the feature to restart analysis from
a specific point in the program) so these test are marked as KNOWNBUG
for now.

Going forward we may re-enable these after fixing the feature, or
possibly do without the restarting entirely.

NlightNFotis committed:
Output C code instead of goto-binary from goto-harness.

Modify goto-harness to output C code that corresponds
to the added harness functions and some global values.
This has been a feature request.

NlightNFotis committed:
Remove unneeded parameter from get_fresh_local_typed_symexpr.

NlightNFotis committed:
Add mode to get_fresh_local_typed_symexpr

NlightNFotis committed:
Add flags to objects created by get_fresh_global_symbol.

NlightNFotis committed:
Add GOTO_HARNESS_PREFIX.

Before this change, we were using CPROVER_PREFIX, which was
causing some symbols to be dropped before the C code output
because of the prefix symbolising CProver internal values.

NlightNFotis committed:
Add mode to param symbol in recursive initialization

NlightNFotis committed:
Make is_initialization_allowed static and change init conditions.

thk123 commented on src/util/type.h:
```diff
@@ -224,4 +224,8 @@ inline type_with_subtypest &to_type_with_subtypes(typet &type)
 
+/// Remove const qualifier from type (if any).
+/// Returns type as is if there is no const qualifier.
+typet remove_const(typet type);
```
:bulb: I guess I'd probably expect this function to recurse... perhaps it should just be a static util function in this place where it is needed?

hannes-steffenhagen-diffblue replied:
I was modelling this after [`std::remove_const`](https://en.cppreference.com/w/cpp/types/remove_cv) which doesn't recurse

chrisr-diffblue replied:
Yeah, I'm not sure why this is a member function rather than a static?

hannes-steffenhagen-diffblue replied:
@chrisr-diffblue it is free standing, not a member. Thomas was suggesting moving it to goto harness where it was used at the time, but this is the sort of thing I could see being useful elsewhere as well and I figured we don't want to have too much proliferation of duplicated utility functions.

chrisr-diffblue replied:
Ah, sorry - I failed to read the diff context properly :-)

thk123 commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -690,12 +690,17 @@ code_blockt recursive_initializationt::build_pointer_constructor(
 
-  const symbol_exprt &local_result =
-    get_fresh_local_typed_symexpr("local_result", type, nullptr_expr);
+  // we want to initialize the pointee as non-const even for pointer to const
```
:pick: Add why to the comment

thk123 commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -797,9 +823,9 @@ code_blockt recursive_initializationt::build_dynamic_array_constructor(
   const optionalt<irep_idt> &lhs_name)
 {
   PRECONDITION(result.type().id() == ID_pointer);
-  const typet &pointer_type = result.type().subtype();
```
:pick: This renaming should have gone in a separate commt

thk123 commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -1008,6 +1008,7 @@ code_blockt recursive_initializationt::build_function_pointer_constructor(
 
   const auto function_pointer_selector =
     get_fresh_local_symexpr("function_pointer_selector");
+  body.add(code_declt{function_pointer_selector});
```
:question: Is there a test for this bug? That is, a test that passes if and only if this line is present

hannes-steffenhagen-diffblue replied:
The existing function pointer tests in goto harness wouldn't really work (anymore) without this.

thk123 commented on src/goto-harness/goto_harness_parse_options.cpp:
```diff
@@ -9,24 +9,58 @@ Author: Diffblue Ltd.
+  return symbols;
+}
+
+static void filter_goto_model(
```
An ideal candidate for a simple unit test

thk123 commented on src/goto-harness/goto_harness_parse_options.cpp:
```diff
@@ -50,6 +84,7 @@ int goto_harness_parse_optionst::doit()
 
   auto factory_options = collect_generate_factory_options();
 
+  register_language(new_ansi_c_language);
```
:no_entry_sign: Could a why comment explaining why this is needed be added here

hannes-steffenhagen-diffblue replied:
@thk123 This snuck in from debugging earlier, it's not needed at all.

thk123 commented on src/goto-harness/goto_harness_parse_options.cpp:
```diff
@@ -87,13 +124,15 @@ int goto_harness_parse_optionst::doit()
+  auto harness_out = std::ofstream{got_harness_config.out_file};
+  dump_c(
+    goto_model.goto_functions,
+    true,
+    true,
+    false,
```
:pick: Pull these out into named local variables so there is some clue as to what they match up to. E.g. 

```C++
const bool use_system_headers = true;
...
dump_c(
  goto_model.goto_functions, 
  use_system_headers, 
...
```

thk123 commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -472,35 +472,54 @@ std::string recursive_initialization_configt::to_string() const
+  auto &fresh_symbol = get_fresh_global_symbol(
+    goto_model.symbol_table,
+    symbol_name,
+    signed_int_type(), // FIXME why always signed_int_type???
```
:question: Noting this FIXME - it does seem odd?

hannes-steffenhagen-diffblue replied:
@thk123 basically this is because of a somewhat odd design decision that was made at some point because someone was adding a bunch of global variables that were all integers apparently... I don't like it, but I considered it out of scope to actually fix this in this PR.

thk123 commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -472,35 +472,54 @@ std::string recursive_initialization_configt::to_string() const
+  irep_idt mode)
 {
+  source_locationt source_location{};
+  source_location.set_file(CPROVER_PREFIX "harness.c");
```
:pick: magic string "harness" should probably be turned into a macro or constexpr

thk123 commented on src/goto-harness/recursive_initialization.h:
```diff
@@ -22,6 +22,7 @@ Author: Diffblue Ltd.
 #include "function_harness_generator_options.h"
 #include "goto_harness_generator.h"
 
+#define GOTO_HARNESS_PREFIX "GOTO_HARNESS"
```
:grey_question: Why not 
```suggestion
#define GOTO_HARNESS_PREFIX CPROVER_PREFIX "_GOTO_HARNESS"
```

hannes-steffenhagen-diffblue replied:
@thk123 anything starting with `CPROVER_PREFIX` is filtered out and ignored by `dump_c`, that's the only reason we added a goto harness specific prefix.

chrisr-diffblue replied:
I would have thought that as this is an "implementation" prefix, it should be `__GOTO_HARNESS` ? (i.e. double underscore at the front) ?

hannes-steffenhagen-diffblue replied:
We could certainly add a `__` in front.

thk123 commented on src/goto-harness/recursive_initialization.h:
```diff
@@ -103,12 +103,14 @@ class recursive_initializationt
   symbol_exprt get_free_function();
 
-  bool is_initialization_allowed(const symbolt &symbol)
+  static bool is_initialization_allowed(const symbolt &symbol)
```
:pick: Great candidate for a unit test

thk123 reviewed (commented):
Yeah this looks mostly good to go to me. Just a few minor :no_entry_sign: comments, and a couple of :question: I'd like answered, but otherwise :+1: and of course figuring out the CI issue!

hannes-steffenhagen-diffblue commented on src/goto-harness/Makefile:
```diff
@@ -16,6 +16,10 @@ OBJ += \
+  ../ansi-c/ansi-c$(LIBEXT) \
+  ../cpp/cpp$(LIBEXT) \
+  ../goto-instrument/dump_c$(OBJEXT) \
+  ../goto-instrument/goto_program2code$(OBJEXT) \
```
these are all needed for `dump_c`

danpoe reviewed (approved):
LGTM

chrisr-diffblue commented on src/util/type.cpp:
```diff
@@ -28,3 +28,9 @@ void type_with_subtypest::move_to_subtypes(typet &type)
   sub.back().swap(type);
 }
+
+typet remove_const(typet type)
```
const argument? Or if this is modifying the argument, then why return ?

hannes-steffenhagen-diffblue replied:
It's returning a copy with the modification. If the argument was const we'd have to make two copies, if it was const reference we'd still have to make a copy so we don't gain anything in either case.

chrisr-diffblue replied:
OK, I like the const just to make it explicit - and of course depending on how much this is called, an extra copy probably isn't the end of the world... But its a nitpick really on my part.

chrisr-diffblue reviewed (commented):
Mostly looks OK, but a couple of queries I'd like answered if possible?

chrisr-diffblue reviewed (approved):
LGTM

smowton commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -774,7 +779,28 @@ code_blockt recursive_initializationt::build_struct_constructor(
+      if(component.type().get_bool(ID_C_constant))
+      {
+        return dereference_exprt{
+          typecast_exprt{address_of_exprt{std::move(member_expr)},
```
Comment that this pile of stuff amounts to `*(X*)&field`, where `field` has type `const X*`

smowton replied:
Also I take it you've tried using braced initialiser style for such things?

smowton commented on src/Makefile:
```diff
@@ -8,8 +8,8 @@ DIRS = analyses \
        goto-diff \
-       goto-harness \
        goto-instrument \
+       goto-harness \
```
Sort order was correct before

smowton reviewed (approved):
Seems ok except the hundred or so tests that get marked KNOWNBUG here -- what was the memory-snapshot thingy that got trashed?

NlightNFotis commented:
@smowton We are tracking the broken tests internally and through github issue https://github.com/diffblue/cbmc/issues/5351. The memory snapshot harness was one of the modes that the `goto-harness` tool supported.
